ghproxy/ghcache/coalesce_test.go

```diff
@@ -92,37 +107,144 @@ func TestRoundTrip(t *testing.T) {
 	}
 }
 
-func runRequest(t *testing.T, rt http.RoundTripper, uri string, immediate bool) {
-	res := make(chan error)
-	run := func() {
-		u, err := url.Parse("http://foo.com" + uri)
-		if err != nil {
-			res <- err
-		}
-		req, err := http.NewRequest(http.MethodGet, u.String(), nil)
-		if err != nil {
-			res <- err
+func TestCacheModeHeader(t *testing.T) {
+	t.Parallel()
+	wg := sync.WaitGroup{}
+	delegate := &testDelegate{
+		hits:            make(map[string]int),
+		beginResponding: sync.NewCond(&sync.Mutex{}),
+	}
+	coalesce := &requestCoalescer{
+		keys:     make(map[string]*responseWaiter),
+		delegate: delegate,
+	}
+
+	checkMode := func(resp *http.Response, expected CacheResponseMode) {
+		mode := CacheResponseMode(resp.Header.Get(CacheModeHeader))
+		if mode != expected {
+			t.Errorf("Expected cache mode %s, but got %s.", string(expected), string(mode))
 		}
-		if immediate {
-			req.Header.Set("test-immediate-response", "true")
+	}
+
+	// Queue an initial request for resource1.
+	// This should eventually return ModeMiss.
+	wg.Add(1)
+	go func() {
+		if resp, err := runRequest(coalesce, "/resource1", false); err != nil {
+			t.Errorf("Failed to run request: %v.", err)
+		} else {
+			checkMode(resp, ModeMiss)
 		}
-		resp, err := rt.RoundTrip(req)
-		if err != nil {
-			res <- err
-		} else if b, err := ioutil.ReadAll(resp.Body); err != nil {
-			res <- err
-		} else if string(b) != "Response" {
-			res <- errors.New("unexpected response value")
+		wg.Done()
+	}()
+	// There is a race here and where sleeps are used below.
+	// We need to wait for the initial request to be made
+	// to the coalescer before letting upstream respond, but we don't have a way
+	// of knowing when the requests has actually started waiting on the
+	// responseWaiter...
+	time.Sleep(time.Second * 3)
```

Comment:
This is long enough to be annoying but likely also long enough not to flake in CI, so works out. We can do a refactor to allow a plug-point for testing later if we need to I guess.

Reply:
What do you mean by "plug-point"?

Reply:
Some interface to allow us to subscribe to the coalesce event in testing